platform/mbed_critical.c

```diff
@@ -51,53 +51,44 @@ bool core_util_is_isr_active(void)
 #endif
 }
 
-MBED_WEAK void core_util_critical_section_enter(void)
+bool core_util_in_critical_section(void)
 {
-    bool interrupts_disabled = !core_util_are_interrupts_enabled();
-    __disable_irq();
-
-    /* Save the interrupt disabled state as it was prior to any nested critical section lock use */
-    if (!interrupt_enable_counter) {
-        critical_interrupts_disabled = interrupts_disabled;
-    }
+    return (critical_section_reentrancy_counter != 0);
+}
 
-    /* If the interrupt_enable_counter overflows or we are in a nested critical section and interrupts
-       are enabled, then something has gone badly wrong thus assert an error.
-    */
-    MBED_ASSERT(interrupt_enable_counter < UINT32_MAX); 
+void core_util_critical_section_enter(void)
+{
 // FIXME
-#ifndef   FEATURE_UVISOR
-    if (interrupt_enable_counter > 0) {
-        MBED_ASSERT(interrupts_disabled);
-    }
+#ifdef FEATURE_UVISOR
+    #warning "core_util_critical_section_enter needs fixing to work from unprivileged code"
 #else
-#warning "core_util_critical_section_enter needs fixing to work from unprivileged code"
+    // If the reentrancy counter overflows something has gone badly wrong.
+    MBED_ASSERT(critical_section_reentrancy_counter < UINT32_MAX);
 #endif /* FEATURE_UVISOR */
-    interrupt_enable_counter++;
+
+    if (critical_section_reentrancy_counter == 0) {
+        hal_critical_section_enter();
```

Comment:
i'm not sure why my previous comment shows as outdated...
this is not thread safe.
There is a global variable `critical_interrupts_enabled` in `hal/mbed_critical_section_api.c` it may get overwritten since time of `critical_section_reentrancy_counter == 0` check can be different from time of `hal_critical_section_enter()` call, as we are not in the critical section yet. This can lead to bugs...
Please consider utilizing exclusive access APIs (LDREX and STREX)

Reply:
> Please consider utilizing exclusive access APIs (LDREX and STREX)

I'm looking at this, but from what I've gathered so far those exclusive access instructions are not supported on Cortex-M0, they were added in ARMv7.

It seems the only way to ensure that access is exclusive is by disabling interrupts, which is problematic as that's what the critical section is for. This is highlighted by the current atomic increment API:

```c++
uint8_t core_util_atomic_incr_u8(uint8_t *valuePtr, uint8_t delta)
{
    uint8_t newValue;
    core_util_critical_section_enter();
    newValue = *valuePtr + delta;
    *valuePtr = newValue;
    core_util_critical_section_exit();
    return newValue;
}
```

Reply:
that is correct for ARMv6.

The previous implementation did not have this, as it disabled irq in the beginning, then access shared variable. @scartmell-arm How this can be made working? there's a potential for race condition

Reply:
There's mixed information on whether it's available. [Can the ARM compiler generate exclusive loads and stores (SWP, LDREX, STREX)](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.faqs/ka4175.html) seems to imply that the instructions are available in ARMv6 (though deprecated), however the CMSIS API doesn't support it, and the ARM reference manual says:

>  ARMv6-M does not support exclusive access instructions such as LDREX, or STREX, or any form of atomic swap instruction. Software must take account of this in multiprocessing environments using shared memory.

Given that, I think the only way is to make this safe is to disable interrupts before accessing the variable, that would require either:

- Disabling the interrupts and re-enabling them at the variable accesses which is not ideal
- Moving the state tracking into the HAL so that the variable can be altered after the platform has disabled interrupts, that's also not ideal as it would put the burden of adding state tracking code onto everyone who implements the platform API. (Though for this it's probably only the 1/2 specific NORDIC boards as well as the default implementation).